cdap-common/src/main/java/io/cdap/cdap/common/internal/remote/DefaultRemoteAuthenticatorProvider.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.common.internal.remote;
+
+import com.google.inject.Inject;
+import com.google.inject.Provider;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.security.spi.authenticator.RemoteAuthenticator;
+
+import javax.annotation.Nullable;
+import javax.inject.Named;
+
+/**
+ * Provides a remote authenticator via the extension loader.
+ */
+public class DefaultRemoteAuthenticatorProvider implements Provider<RemoteAuthenticator> {
+
+  public static final String OVERRIDE_AUTHENTICATOR_CONFIG = "OverrideRemoteAuthenticatorConfig";
+
+  private final String defaultAuthenticatorName;
+  private final RemoteAuthenticatorExtensionLoader remoteAuthenticatorExtensionLoader;
+  private GceRemoteAuthenticator gceRemoteAuthenticator;
+
+  @Inject
+  DefaultRemoteAuthenticatorProvider(CConfiguration cConf,
+                                     @Named(OVERRIDE_AUTHENTICATOR_CONFIG) @Nullable String overrideAuthenticatorConfig,
+                                     RemoteAuthenticatorExtensionLoader remoteAuthenticatorExtensionLoader) {
+    // Use the override authenticator for subcomponents if it's set, but default to overall system wide configuration
+    // if no override is set.
+    if (overrideAuthenticatorConfig != null && cConf.get(overrideAuthenticatorConfig) != null) {
+      this.defaultAuthenticatorName = cConf.get(overrideAuthenticatorConfig);
+    } else {
+      this.defaultAuthenticatorName = cConf.get(Constants.RemoteAuthenticator.REMOTE_AUTHENTICATOR_NAME);;
+    }
+    this.remoteAuthenticatorExtensionLoader = remoteAuthenticatorExtensionLoader;
+    this.gceRemoteAuthenticator = null;
+  }
+
+  /**
+   * Retrieves the current {@link RemoteAuthenticator} from the extension loader using the authenticator name or
+   * {@code null} if there is no current authenticator available.
+   */
+  @Nullable
+  @Override
+  public RemoteAuthenticator get() {
+    if (defaultAuthenticatorName == null) {
+      return null;
```

Comment:
Avoid using `null` for guice injection. Return a no-op authenticator instead.

Reply:
Done

Reply:
Removed all Nullable annotations as well.